Sources/xcodes/main.swift

```diff
@@ -20,17 +21,24 @@ enum Error: Swift.Error {
     case unavailableVersion(Version)
 }
 
-func loginIfNeeded() -> Promise<Void> {
+func loginIfNeeded(withUsername existingUsername: String? = nil) -> Promise<Void> {
     return firstly { () -> Promise<Void> in
         return manager.client.validateSession()
     }
     .recover { error -> Promise<Void> in
         guard
-            let username = findUsername() ?? readLine(prompt: "Apple ID: "),
+            let username = existingUsername ?? findUsername() ?? readLine(prompt: "Apple ID: "),
             let password = findPassword(withUsername: username) ?? readSecureLine(prompt: "Apple ID Password: ")
         else { throw Error.missingUsernameOrPassword }
 
-        return login(username, password: password)
+        return firstly { () -> Promise<Void> in
+            login(username, password: password)
+        }
+        .recover { error -> Promise<Void> in
+            print(error.legibleLocalizedDescription)
+            print("Try entering your password again")
+            return loginIfNeeded(withUsername: username)
```

Comment:
When I was testing to make sure that the keychain item was only removed when there was a password issue, I found that if I turned off my internet connection it would get stuck in a retry loop that never prompted for a password. I think it didn't prompt because the password was still in the keychain (and returned in `findPassword` above) but it'll never succeed because the internet isn't available. Seems like the same check for a password failure should happen here too.

I'm sorry that I didn't find this sooner during review. 😞 I think this'll be the last thing before merging.

```suggestion
            if case Client.Error.invalidUsernameOrPassword = error {
                print("Try entering your password again")
                return loginIfNeeded(withUsername: username)
            }
            else {
                return Promise(error: error)
            }
```